cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -0,0 +1,1588 @@
+                    f"'adata.obsm['{key}']' is {type(value)}')"
+                )
+                continue
+
+            # Embeddings to be shown in cellxgene explorer
+            if key.startswith("X_"):
+                obsm_with_x_prefix += 1
+
+                if len(value.shape) < 2:
+                    self.errors.append(
+                        f"All embeddings must have as many rows as cells, and at least two columns."
+                        f"'adata.obsm['{key}']' has shape of '{value.shape}'"
+                    )
+                else:
+                    if value.shape[0] != self.adata.n_obs or value.shape[1] < 2:
+                        self.errors.append(
+                            f"All embeddings must have as many rows as cells, and at least two columns."
+                            f"'adata.obsm['{key}']' has shape of '{value.shape}'"
+                        )
+
+        if obsm_with_x_prefix == 0:
+            self.errors.append(
+                "At least one embedding in 'obsm' has to have a key with an 'X_' prefix."
+            )
+
+    def _are_children_of(
+        self, component: str, column: str, ontology_name: str, ancestors: List[str]
+    ) -> bool:
+
+        """
+        Checks if elements in the specified column of the component (e.g. 'assay_ontology_term_id' of 'adata.obs') are
+        children of the given ancestors.
+
+        Ancestors checks are inclusive, meaning that a value it's its own ancestor as well.
+
+        :param str component: the name of the component that's been checked.
+        :param str column: Column in the component to check
+        :param str ontology_name: Name of the ontology (e.g. "EFO")
+        :param List[str] ancestors: List of ancestors
+
+        :rtype bool
+        :return True if any value in column is children of any ancestor.
+        """
+
+        curies = getattr(getattr(self.adata, component), column)
+        curies = curies.drop_duplicates()
+
+        for curie in curies:
+            if ONTOLOGY_CHECKER.is_valid_term_id(ontology_name, curie):
+                curie_ancestors = ONTOLOGY_CHECKER.get_term_ancestors(
+                    ontology_name, curie
+                )
+                curie_ancestors.add(curie)
+                if bool(curie_ancestors & set(ancestors)):
+                    return True
+
+        return False
+
+    def _get_raw_x(self) -> Union[ndarray, sparse.csc_matrix, sparse.csr_matrix]:
+
+        """
+        gets raw x (best guess, i.e. not guarantee it's actually raw)
+        """
+
+        if self.adata.raw:
+            return self.adata.raw.X
+        else:
+            return self.adata.X
+
+    def _get_raw_x_loc(self) -> str:
+
+        """
+        gets raw x location (best guess, i.e. not guarantee it's actually raw)
+        """
+
+        if self.adata.raw:
+            return "raw.X"
+        else:
+            return "X"
+
+    def _is_raw(self, n_values_to_check: int = 5000, force: bool = False) -> bool:
```

Comment:
any rationale to the 5000 value?   if checking more than this is a performance concern, worth a comment, and could rename to `max_values_to_check`

Reply:
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203